### PR TITLE
Use aws/rds as default KMS key

### PR DESCRIPTION
**Pull request by alexjurkiewicz**
If the user enables storage or performance insights encryption without
specifying a KMS key, use the AWS-managed aws/rds one.

There's one thing I'm checking. Does the `aws/rds` key always exist in AWS accounts, or is it only created when needed? If the latter, loading it as a data source could fail in new accounts, leading to a chicken and egg problem with this module attempting to deploy a cluster in a new account. I've asked AWS support for clarification.

### CODE REVIEW

**Review by Gowiem — APPROVED** [review]
I like it 👍 

I'm pretty sure the `aws/*` KMS keys are always around in an account, but we'll wait to merge until you hear back from AWS support to be 100% sure.

**Comment by alexjurkiewicz** [comment]
Unfortunately, AWS Support confirmed the key is created on-demand:
https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/Overview.Encryption.html#Overview.Encryption.Enabling

> **When you create an encrypted DB instance**, you can choose a customer managed CMK or the AWS managed CMK for Amazon RDS to encrypt your DB instance. **If you don't specify the key identifier for a customer managed CMK**, Amazon RDS uses the AWS managed CMK for your new DB instance. **Amazon RDS creates an AWS managed CMK for Amazon RDS for your AWS account.** Your AWS account has a different AWS managed CMK for Amazon RDS for each AWS Region. 

So, it's probably not OK for this module to **require** the aws/rds key. Instead, I will change the logic so this key is only looked up if the user doesn't specify their own keys (a good optimisation anyway), and document the restriction. I'll also investigate if there's a simpler way to create this key than "create an instance with encryption enabled but no key specified".

**Comment by alexjurkiewicz** [comment]
I've updated this PR so that the aws/rds key is only loaded with a data source if required, and mentioned the "must already exist" restriction in the docs.

Unfortunately, I don't think it's possible to have this module create the key for you. You can create a db instance with `storage_enabled = true` and `kms_key_arn = null`, but Terraform will detect `kms_key_arn` as changed on subsequent runs.

**Comment by alexjurkiewicz** [comment]
ping 🙏

**Comment by Gowiem** [comment]
@alexjurkiewicz discussing this one internally with the team. Now that we know those AWS KMS Keys are not around by default in new accounts... I'm wondering if we should just use `cloudposse/kms-key/aws` in the cases where no key is passed but `storage_encrypted` is true. Let me know your thoughts and I'll also discuss with the team. 

Thanks!

**Comment by alexjurkiewicz** [comment]
That makes sense. A KMS key (without rotation) is $1/mo, which is nothing compared to the cost of an RDS cluster.

**Comment by alexjurkiewicz** [comment]
d1a677f:

>  Create KMS keys if needed
> 
> Rather than rely on aws/rds, which isn't guaranteed to exist (it's
> created in each account/region when you create an RDS resource with
> encryption and don't specify an explicit KMS key).
> 
> Create separate keys for storage and performance insights, adding a
> hardcoded policy to both (copied from aws/rds). Output the key ARNs as
> well.
> 
> Finally, split KMS logic out to a separate file, because the policy is
> so large it makes main.tf hard to read otherwise.

File: kms.tf [hunk 1]
```diff
@@ -0,0 +1,87 @@
+locals {
+  storage_create_kms_key              = var.storage_encrypted && var.kms_key_arn == "" ? true : false
```
Comment (nitrocode) [inline kms.tf:2]:
```suggestion
  storage_create_kms_key              = var.storage_encrypted && var.kms_key_arn == ""
```

File: kms.tf [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+  storage_kms_key_arn                 = local.storage_create_kms_key ? module.storage_kms_key[1].key_arn : var.kms_key_arn
+  performance_insights_create_kms_key = var.performance_insights_enabled && var.performance_insights_kms_key_id == "" ? true : false
```
Comment (nitrocode) [inline kms.tf:4]:
```suggestion
  performance_insights_create_kms_key = var.performance_insights_enabled && var.performance_insights_kms_key_id == ""
```

File: kms.tf [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+  performance_insights_kms_key_id     = local.performance_insights_create_kms_key ? module.performance_insights_kms_key[1].key_arn : var.performance_insights_kms_key_id
+
+  # This policy is copied from aws/rds
+  kms_policy = <<-EOT
```
Comment (nitrocode) [inline kms.tf:8]: This should be a data aws iam policy document. You can convert it using iam policy json to terraform tool https://github.com/flosell/iam-policy-json-to-terraform

File: kms.tf [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+                      "kms:CallerAccount": "${data.aws_caller_identity.current.account_id}"
+                  }
+              }
+          },
+          {
+              "Sid": "Allow direct access to key metadata to the account",
+              "Effect": "Allow",
+              "Principal": {
+                  "AWS": "arn:aws:iam::${data.aws_caller_identity.current.account_id}:root"
+              },
```
Comment (nitrocode) [inline kms.tf:50]: Use the data partition here instead of aws so govcloud can make use of this

File: kms.tf [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+data "aws_region" "current" {}
+data "aws_caller_identity" "current" {}
+data "aws_partition" "current" {}
+
+module "storage_kms_key" {
+  for_each            = local.storage_create_kms_key ? [1] : []
+  source              = "cloudposse/kms-key/aws"
+  version             = "~> 0.9"
+  context             = module.this.id
```
Comment (nitrocode) [inline kms.tf:72]:
```suggestion
  context             = module.this.context
```

File: kms.tf [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+  description         = "RDS storage encryption for ${module.this.id}"
+  enable_key_rotation = false
+  policy              = local.kms_policy
+}
+
+module "performance_insights_kms_key" {
+  for_each            = local.storage_create_kms_key ? [1] : []
+  source              = "cloudposse/kms-key/aws"
+  version             = "~> 0.9"
+  context             = module.this.id
```
Comment (nitrocode) [inline kms.tf:82]:
```suggestion
  context             = module.this.context
```

**Comment by Gowiem** [comment]
@alexjurkiewicz friendly ping on this one -- Still want to move this forward or should we close it?

**Comment by hans-d** [comment]
closing

**Comment by hans-d** [comment]
closing

**Pull request closed without merging**